TESTS/mbed_hal/sai/main.cpp

```diff
@@ -0,0 +1,523 @@
+            const loopback_test_tatus_t result = sai_loopback_communication_test(&test_sai_formats[f_idx], test_sample_rates[sr_idx]);
+
+            /* First `required_formats_count` formats are obligate, so we expect that
+             * communication tests passes. Others formats may be supported, so test can pass or be skipped.
+             */
+            sprintf(message, "format idx: %lu, sample rate idx: %lu", f_idx, sr_idx);
+            if (f_idx < required_formats_count) {
+                TEST_ASSERT_EQUAL_MESSAGE(LOOPBACK_TEST_OK, result, message);
+            } else {
+                TEST_ASSERT_NOT_EQUAL_MESSAGE(LOOPBACK_TEST_FAILURE, result, message);
+            }
+        }
+    }
+}
+
+/* Test that sai_init() returns `SAI_RESULT_ALREADY_INITIALIZED` if SAI is to be re-initialised. */
+void sai_reinit_test()
+{
+    sai_t sai_obj = { };
+    sai_init_t sai_init_data = { };
+
+    sai_config_create(&sai_init_data, false, SAI_DEFAULT_SAMPLE_RATE, (sai_format_t*)&sai_mode_i2s16w32);
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_OK, sai_init(&sai_obj, &sai_init_data));
+
+    /* Change config (format) and try to re-init. */
+    sai_config_create(&sai_init_data, false, SAI_DEFAULT_SAMPLE_RATE, (sai_format_t*)&sai_mode_i2s16);
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_ALREADY_INITIALIZED, sai_init(&sai_obj, &sai_init_data));
+
+    sai_free(&sai_obj);
+}
+
+/* Test that sai_init() returns `SAI_RESULT_INVALID_PARAM` if at least one of the given parameters is
+ *   undefined (NULL). */
+void sai_init_invalid_param_test()
+{
+    sai_t sai_obj;
+    sai_init_t sai_init_data = { };
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_INVALID_PARAM, sai_init(NULL, &sai_init_data));
+    TEST_ASSERT_EQUAL(SAI_RESULT_INVALID_PARAM, sai_init(&sai_obj, NULL));
+}
+
+/* Test that sai_init() returns `SAI_RESULT_CONFIG_UNSUPPORTED` if the device can never support
+ * the specified configuration. */
+void sai_init_unsupported_config_test()
+{
+    sai_t sai_obj = { };
+    sai_init_t sai_init_data = { };
+
+    /* Create invalid configuration: invalid sample rate. */
+    sai_config_create(&sai_init_data, false, 0, (sai_format_t*)&sai_mode_i2s16w32);
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_CONFIG_UNSUPPORTED, sai_init(&sai_obj, &sai_init_data));
+
+    /* Create invalid configuration: invalid pin. */
+    sai_config_create(&sai_init_data, false, SAI_DEFAULT_SAMPLE_RATE, (sai_format_t*)&sai_mode_i2s16w32);
+    sai_init_data.mclk = LED1;
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_CONFIG_UNSUPPORTED, sai_init(&sai_obj, &sai_init_data));
+
+    /* Create invalid configuration: invalid format (word length inconsistent with data length). */
+    sai_config_create(&sai_init_data, false, SAI_DEFAULT_SAMPLE_RATE, (sai_format_t*)&sai_mode_i2s16w32);
+    sai_init_data.format.word_length = 16;
+    sai_init_data.format.data_length = 32;
+
+    TEST_ASSERT_EQUAL(SAI_RESULT_CONFIG_UNSUPPORTED, sai_init(&sai_obj, &sai_init_data));
```

Comment:
As word length shorter than the data length should probably be treated as an `SAI_RESULT_INVALID_PARAM` instead.